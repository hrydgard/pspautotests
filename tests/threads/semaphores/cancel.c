#include "shared.h"

SETUP_SCHED_TEST;

#define CANCEL_TEST(title, sema, count) { \
	int result = sceKernelCancelSema(sema, count, NULL); \
	if (result == 0) { \
		schedf("%s: OK\n", title); \
	} else { \
		schedf("%s: Failed (%X)\n", title, result); \
	} \
	PRINT_SEMAPHORE(sema); \
}

#define CANCEL_TEST_WITH_WAIT(title, sema, count) { \
	int waitThreads = 99; \
	int result = sceKernelCancelSema(sema, count, &waitThreads); \
	if (result == 0) { \
		schedf("%s: OK (%d waiting)\n", title, waitThreads); \
	} else { \
		schedf("%s: Failed (%X, %d waiting)\n", title, result, waitThreads); \
	} \
	PRINT_SEMAPHORE(sema); \
}

void foo() {
}

int main(int argc, char **argv) {
	SceUID sema = sceKernelCreateSema("cancel", 0, 0, 1, NULL);

	CANCEL_TEST("Normal", sema, 1);
	CANCEL_TEST("Greater than max", sema, 3);
	CANCEL_TEST("Zero", sema, 0);
	CANCEL_TEST("Negative -3", sema, -3);
	CANCEL_TEST("Negative -1", sema, -1);

	CANCEL_TEST_WITH_WAIT("Normal", sema, 1);
	CANCEL_TEST_WITH_WAIT("Greater than max", sema, 3);
	CANCEL_TEST_WITH_WAIT("Zero", sema, 0);
	CANCEL_TEST_WITH_WAIT("Negative -3", sema, -3);
	CANCEL_TEST_WITH_WAIT("Negative -1", sema, -1);

	sceKernelDeleteSema(sema);

	TWO_STEP_SCHED_TEST("Cancel waited 0 then 1", 0, 1,
		result = 0;
		CANCEL_TEST_WITH_WAIT("To 0", sema1, 0);
	,
		CANCEL_TEST_WITH_WAIT("To 1", sema1, 1);
	);

	BASIC_SCHED_TEST("NULL",
		result = sceKernelCancelSema(0, 0, NULL);
	);
	BASIC_SCHED_TEST("Cancel bad count",
		result = sceKernelCancelSema(sema2, 0, NULL);
	);
	BASIC_SCHED_TEST("Cancel other",
		result = sceKernelCancelSema(sema2, 1, NULL);
	);
	BASIC_SCHED_TEST("Cancel waited",
		result = sceKernelCancelSema(sema1, 1, NULL);
	);
	BASIC_SCHED_TEST("Cancel other with threads",
		int n;
		result = sceKernelCancelSema(sema2, 1, &n);
	);
	BASIC_SCHED_TEST("Cancel waited with threads",
		int n;
		result = sceKernelCancelSema(sema1, 1, &n);
	);

	CANCEL_TEST("NULL", 0, 0);
	CANCEL_TEST("Invalid", 0xDEADBEEF, 0);
	CANCEL_TEST("Deleted", sema, 0);

	return 0;
}