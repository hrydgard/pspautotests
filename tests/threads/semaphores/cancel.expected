Normal: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Greater than max: Failed (800201BD)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Zero: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -3: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -1: OK
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Normal: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Greater than max: Failed (800201BD, 99 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=1,max=1,wait=0)
Zero: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -3: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Negative -1: OK (0 waiting)
Sema: OK (size=56,name='cancel',attr=0,init=0,cur=0,max=1,wait=0)
Cancel waited 0 then 1: A1B1C2D1To 0: OK (1 waiting)
Sema: OK (size=56,name='schedTest1',attr=0,init=0,cur=0,max=1,wait=0)
E2To 1: OK (0 waiting)
Sema: OK (size=56,name='schedTest1',attr=0,init=0,cur=1,max=1,wait=0)
F1 (thread=800201A9, main=00000000)
NULL: A1B1C2E1D1F2 (thread=800201B5, main=80020199)
Cancel bad count: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Cancel other: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Cancel waited: A1B1C2D1E2F1 (thread=800201A9, main=00000000)
Cancel other with threads: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Cancel waited with threads: A1B1C2D1E2F1 (thread=800201A9, main=00000000)
NULL: Failed (80020199)
Sema: Failed (00000000)
Invalid: Failed (80020199)
Sema: Failed (DEADBEEF)
Deleted: Failed (80020199)
Sema: Invalid (80020199)
